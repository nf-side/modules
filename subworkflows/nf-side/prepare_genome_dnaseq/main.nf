include { BWAMEM2_INDEX                  } from '../../../modules/nf-core/bwamem2/index'
include { BWA_INDEX as BWAMEM1_INDEX     } from '../../../modules/nf-core/bwa/index'
include { DRAGMAP_HASHTABLE              } from '../../../modules/nf-core/dragmap/hashtable'
include { GATK4_CREATESEQUENCEDICTIONARY } from '../../../modules/nf-core/gatk4/createsequencedictionary'
include { GAWK as BUILD_INTERVALS        } from '../../../modules/nf-core/gawk'
include { MSISENSORPRO_SCAN              } from '../../../modules/nf-core/msisensorpro/scan'
include { SAMTOOLS_FAIDX                 } from '../../../modules/nf-core/samtools/faidx'
include { TABIX_BGZIPTABIX               } from '../../../modules/nf-core/tabix/bgziptabix'
include { TABIX_TABIX                    } from '../../../modules/nf-core/tabix/tabix'

workflow PREPARE_GENOME_DNASEQ {
    take:
    fasta                        // channel: [meta, fasta]
    fasta_fai                    // channel: [meta, fasta_fai]
    vcf                          // channel: [meta, vcf]
    run_bwamem1                  // boolean: true/false
    run_bwamem2                  // boolean: true/false
    run_createsequencedictionary // boolean: true/false
    run_dragmap                  // boolean: true/false
    run_faidx                    // boolean: true/false
    run_intervals                // boolean: true/false
    run_msisensorpro             // boolean: true/false
    run_tabix                    // boolean: true/false

    main:
    bwamem1_index = Channel.empty()
    bwamem2_index = Channel.empty()
    dragmap_hashmap = Channel.empty()
    fasta_dict = Channel.empty()
    intervals_bed = Channel.empty()
    msisensorpro_list = Channel.empty()
    vcf_gz = Channel.empty()
    vcf_tbi = Channel.empty()

    versions = Channel.empty()

    if (run_bwamem1) {
        BWAMEM1_INDEX(fasta)

        bwamem1_index = BWAMEM1_INDEX.out.index
        versions = versions.mix(BWAMEM1_INDEX.out.versions)
    }

    if (run_bwamem2) {
        BWAMEM2_INDEX(fasta)

        bwamem2_index = BWAMEM2_INDEX.out.index
        versions = versions.mix(BWAMEM2_INDEX.out.versions)
    }

    if (run_dragmap) {
        DRAGMAP_HASHTABLE(fasta)

        dragmap_hashmap = DRAGMAP_HASHTABLE.out.hashmap
        versions = versions.mix(DRAGMAP_HASHTABLE.out.versions)
    }

    if (run_createsequencedictionary) {
        GATK4_CREATESEQUENCEDICTIONARY(fasta)

        fasta_dict = GATK4_CREATESEQUENCEDICTIONARY.out.dict
        versions = versions.mix(GATK4_CREATESEQUENCEDICTIONARY.out.versions)
    }

    if (run_faidx || run_intervals) {

        if (run_faidx) {
            // Do not generate sizes for DNAseq
            generate_sizes = false

            SAMTOOLS_FAIDX(fasta.map { meta, fasta_ -> [meta, fasta_, []] }, generate_sizes)

            fasta_fai = fasta_fai.mix(SAMTOOLS_FAIDX.out.fai)
            versions = versions.mix(SAMTOOLS_FAIDX.out.versions)
        }

        if (run_intervals) {
            BUILD_INTERVALS(fasta_fai, [], false)
            intervals_bed = BUILD_INTERVALS.out.output
            versions = versions.mix(BUILD_INTERVALS.out.versions)
        }
    }

    if (run_msisensorpro) {
        MSISENSORPRO_SCAN(fasta)

        msisensorpro_list = MSISENSORPRO_SCAN.out.list
        versions = versions.mix(MSISENSORPRO_SCAN.out.versions)
    }

    if (run_tabix) {
        vcf_to_index = vcf.branch { _meta, vcf_ ->
            vcf_gz: vcf_.toString().endsWith('.gz')
            vcf: true
        }

        TABIX_BGZIPTABIX(vcf_to_index.vcf)
        TABIX_TABIX(vcf_to_index.vcf_gz)

        vcf_gz = TABIX_BGZIPTABIX.out.gz_tbi.map { meta, vcf_gz_, _vcf_tbi -> [meta, vcf_gz_] }
        vcf_tbi = TABIX_TABIX.out.tbi.mix(TABIX_BGZIPTABIX.out.gz_tbi.map { meta, _vcf_gz, vcf_tbi_ -> [meta, vcf_tbi_] })

        versions = versions.mix(TABIX_BGZIPTABIX.out.versions)
        versions = versions.mix(TABIX_TABIX.out.versions)
    }

    emit:
    bwamem1_index     // channel: [meta, BWAmemIndex/]
    bwamem2_index     // channel: [meta, BWAmem2memIndex/]
    dragmap_hashmap   // channel: [meta, DragmapHashtable/]
    fasta_dict        // channel: [meta, *.fa(sta).dict]
    fasta_fai         // channel: [meta, *.fa(sta).fai]
    intervals_bed     // channel: [meta, *.bed]
    msisensorpro_list // channel: [meta, *.list]
    vcf_gz            // channel: [meta, *.vcf.gz]
    vcf_tbi           // channel: [meta, *.vcf.gz.tbi]
    versions          // channel: [versions.yml]
}
