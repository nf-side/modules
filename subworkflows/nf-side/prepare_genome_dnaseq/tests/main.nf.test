nextflow_workflow {

    name "Test PREPARE_GENOME_DNASEQ"
    script "../main.nf"
    workflow "PREPARE_GENOME_DNASEQ"

    test(".fasta/.vcf/.vcf.gz - all tools") {

        when {
            params {
            }
            workflow {
                """
                run_bwamem1 = true
                run_bwamem2 = true
                run_createsequencedictionary = true
                run_dragmap = true
                run_faidx = true
                run_intervals = true
                run_msisensorpro = true
                run_tabix = true

                input[0] = Channel.of([
                    [id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                input[1] = Channel.empty()
                input[2] = Channel.of([
                    [id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz', checkIfExists: true)
                ],[
                    [id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/vcf/genmod_compound.vcf', checkIfExists: true)
                ])
                input[3] = run_bwamem1
                input[4] = run_bwamem2
                input[5] = run_createsequencedictionary
                input[6] = run_dragmap
                input[7] = run_faidx
                input[8] = run_intervals
                input[9] = run_msisensorpro
                input[10] = run_tabix
                """
            }
        }

        then {
            def all_files_temp = file(workflow.out.dragmap_hashmap.get(0).get(1)).listFiles().sort()
            def dragmap_hashmap = all_files_temp.collect { file -> file.name }
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.bwamem1_index,
                    workflow.out.bwamem2_index,
                    workflow.out.dragmap_hashmap.collect { [it[0], file(it[1]).name + "/", dragmap_hashmap] },
                    workflow.out.fasta_dict,
                    workflow.out.fasta_fai,
                    workflow.out.intervals_bed,
                    workflow.out.msisensorpro_list,
                    workflow.out.vcf_gz.collect { [it[0], "${file(it[1]).name},variantsMD5:${path(it[1]).vcf.variantsMD5}"] },
                    workflow.out.vcf_tbi.collect { [it[0], file(it[1]).name] },
                    workflow.out.versions
                ).match()
                }
            )
        }
    }
}
