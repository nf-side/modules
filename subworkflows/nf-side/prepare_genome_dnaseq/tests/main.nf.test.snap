{
    ".fasta/.vcf/.vcf.gz - all tools": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "genome.amb:md5,1891c1de381b3a96d4e72f590fde20c1",
                        "genome.ann:md5,2df4aa2d7580639fa0fcdbcad5e2e969",
                        "genome.bwt:md5,815eded87e4cb6b0f1daab5c4d6e30af",
                        "genome.pac:md5,8569fbdb2c98c6fb16dfa73d8eacb070",
                        "genome.sa:md5,e7cff62b919448a3a3d0fe4aaf427594"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "genome.fasta.0123:md5,d73300d44f733bcdb7c988fc3ff3e3e9",
                        "genome.fasta.amb:md5,1891c1de381b3a96d4e72f590fde20c1",
                        "genome.fasta.ann:md5,2df4aa2d7580639fa0fcdbcad5e2e969",
                        "genome.fasta.bwt.2bit.64:md5,cd4bdf496eab05228a50c45ee43c1ed0",
                        "genome.fasta.pac:md5,8569fbdb2c98c6fb16dfa73d8eacb070"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "dragmap/",
                    [
                        "hash_table.cfg",
                        "hash_table.cfg.bin",
                        "hash_table.cmp",
                        "hash_table_stats.txt",
                        "ref_index.bin",
                        "reference.bin",
                        "repeat_mask.bin",
                        "str_table.bin"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "genome.dict:md5,2433fe2ba31257337bf4c4bd4cb8da15"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "genome.fasta.fai:md5,3520cd30e1b100e55f578db9c855f685"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.bed:md5,87a15eb9c2ff20ccd5cd8735a28708f7"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.msisensor_scan.list:md5,a7886e7a56a1d7e3be6b4496753fd743"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.vcf.gz,variantsMD5:4af4335dc9afaa1d0981cb037da97b59"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "dbsnp_146.hg38.vcf.gz.tbi"
                ],
                [
                    {
                        "id": "test"
                    },
                    "test.vcf.gz.tbi"
                ]
            ],
            [
                "versions.yml:md5,1a261512dbad505711e60f42b540b715",
                "versions.yml:md5,2fd70a9efdd32615abdd8f7782dc96ff",
                "versions.yml:md5,559faafdcd1e8cfbe4508196bff7117b",
                "versions.yml:md5,757894083e1c8c3e6facc452a706be35",
                "versions.yml:md5,9921c6dc6cad5895445bd3813814685a",
                "versions.yml:md5,9b316374534f9598679f3c714477e1b2",
                "versions.yml:md5,b5f8d54c5dcdf08109279ea5ba00c243",
                "versions.yml:md5,c4dafb726ca57f4b5deda9010c825380",
                "versions.yml:md5,cc25e0d0b6f8914bcfa47ad9c5dfde0b"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.01.0"
        },
        "timestamp": "2025-03-07T13:14:25.14497332"
    }
}